internal/endorser/endorser.go

```diff
@@ -0,0 +1,169 @@
+// provided. The endorsement statement is generated only if the provenance statements are valid.
+// Each provenanceURI must either specify a local file (using the `file` scheme), or a remote file
+// (using the `http/https` scheme).
+func GenerateEndorsement(binaryHash string, validityDuration amber.ClaimValidity, provenanceURIs []string) (*intoto.Statement, error) {
+	verifiedProvenances, err := loadAndVerifyProvenances(provenanceURIs, binaryHash)
+	if err != nil {
+		return nil, fmt.Errorf("could not load provenances: %v", err)
+	}
+
+	return amber.GenerateEndorsementStatement(validityDuration, *verifiedProvenances), nil
+}
+
+// Returns an instance of amber.VerifiedProvenanceSet, containing metadata about a set of verified
+// provenances, or an error. An error is returned if any of the following conditions is met:
+// (1) The list of provenanceURIs is empty,
+// (2) Any of the provenances cannot be loaded (e.g., invalid URI),
+// (3) Any of the provenances is invalid (see verifyProvenances for details on validity),
+// (4) Provenances do not match (e.g., have different binary names).
+func loadAndVerifyProvenances(provenanceURIs []string, binaryHash string) (*amber.VerifiedProvenanceSet, error) {
+	if len(provenanceURIs) == 0 {
+		return nil, fmt.Errorf("at least one provenance file must be provided")
+	}
+
+	// load provenances from URIs
+	provenances := make([]amber.ValidatedProvenance, 0, len(provenanceURIs))
+	provenancesData := make([]amber.ProvenanceData, 0, len(provenanceURIs))
+	for _, uri := range provenanceURIs {
+		provenanceBytes, err := getProvenanceBytes(uri)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't load the provenance file from %s: %v", uri, err)
+		}
+		provenance, err := amber.ParseProvenanceData(provenanceBytes)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't parse bytes from %s into a provenance statement: %v", uri, err)
+		}
+		sum256 := sha256.Sum256(provenanceBytes)
+		provenances = append(provenances, *provenance)
+		provenancesData = append(provenancesData, amber.ProvenanceData{
+			URI:        uri,
+			SHA256Hash: hex.EncodeToString(sum256[:]),
+		})
+	}
+
+	if err := verifyProvenances(provenances, binaryHash); err != nil {
+		return nil, fmt.Errorf("verification of provenances failed: %v", err)
+	}
+
+	verifiedProvenances := amber.VerifiedProvenanceSet{
+		BinaryName:  provenances[0].GetBinaryName(),
+		BinaryHash:  binaryHash,
+		Provenances: provenancesData,
+	}
+
+	return &verifiedProvenances, nil
+}
+
+// verifyProvenances verifies that the given list of provenances are consistent and have the given
+// binaryHash as the subject. Provenances are consistent if they all have the same binary name and
+// binary hash. An error is returned if any of the conditions above are violated.
+// TODO(b/222440937): Document any additional checks.
+func verifyProvenances(provenances []amber.ValidatedProvenance, binaryHash string) error {
```

Comment:
I'm not sure I fully understand why verifying provenances would be necessary for an endorsement.

In practice in most cases I would expect that before endorsing a binary, the provenance has already been checked (e.g. when checking in the file to google3), plus ideally more information about the quality of the binary; but just asserting that there exists any provenance for the binary seems a bit arbitrary, unless something else about the provenance is also checked. For example, anyone can produce a valid provenance of any binary, just by checking in to any github repository and running an "identity" builder that only emits that binary.

But probably you are thinking of something more interesting to check further down the line?

Or is the idea that as part of reviewing the binary for release, all the evidence considered is also forwarded to the endorsement, for future reference, in case someone wonders how the person endorsing the binary arrived to the conclusion that the binary was fine?

In any case, it does not hurt to have it here, but I'm worried about duplicating a subset of this logic, because that will also have to live somewhere else in order to ensure that binaries in google3 have the correct provenance (even if they never get endorsed). So perhaps a way of addressing my point is for this logic to be in the verifier instead of the endorser? In other words, we should not have to actually release a binary in order to find out that any of these preconditions related to provenance failed. Or is there another mechanism that will enforce that?

Reply:
> ...but just asserting that there exists any provenance for the binary seems a bit arbitrary...

That is not the idea. The goal is to make sure that provenances are valid before generating the endorsement. You mentioned that "the provenance has already been checked". How would the endorser know that? The ideas is that the endorser will itself run all the checks before generating the endorsement. 

> But probably you are thinking of something more interesting to check further down the line?

Yes. This is the endorser with the simplest set of checks. @mariaschett is designing a more realistic verifier.

> Or is the idea that as part of reviewing the binary for release, all the evidence considered is also forwarded to the endorsement, for future reference, in case someone wonders how the person endorsing the binary arrived to the conclusion that the binary was fine?

Yes. That is the idea, and all the input will be recorded as evidence in the endorsement statement. 

> In any case, it does not hurt to have it here, but I'm worried about duplicating a subset of this logic, because that will also have to live somewhere else in order to ensure that binaries in google3 have the correct provenance (even if they never get endorsed). So perhaps a way of addressing my point is for this logic to be in the verifier instead of the endorser? 

That is basically the plan. @mariaschett is designing a verifier **library** that will be used both here and in `cmd/verifier`. The `b/222440937` TODOs are exactly where the code should be replaced with calls to functionality from the verifier. 

> In other words, we should not have to actually release a binary in order to find out that any of these preconditions related to provenance failed. Or is there another mechanism that will enforce that?

The verifier tool can be used to perform any necessary checks without actually generating an endorsement statement. We could alternatively merge the two tools and add a dry-run option to the endorser to simulate the verifier. But I prefer having separate tools.

Reply:
> The verifier tool can be used to perform any necessary checks without actually generating an endorsement statement. We could alternatively merge the two tools and add a dry-run option to the endorser to simulate the verifier. But I prefer having separate tools.

Me too, that's why I was expecting this logic to be in the verifier, not in the endorser. But maybe the move to a verifier library you mentioned above will address that for me.